nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/PutMongoBulkOperations.java

```diff
@@ -0,0 +1,304 @@
+        final WriteConcern writeConcern = clientService.getWriteConcern();
+
+        ClientSession clientSession = null;
+        try {
+            final MongoCollection<Document> collection = getCollection(context, flowFile).withWriteConcern(writeConcern);
+            // Read the contents of the FlowFile into a byte array
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+            session.exportTo(flowFile, os);
+
+            // parse
+            final BsonArray updateItems = BsonArray.parse(os.toString(charset));
+
+            List<WriteModel<Document>> updateModels = new ArrayList<>();
+            for (Object item : updateItems) {
+                final BsonDocument updateItem = (BsonDocument) item;
+                if (updateItem.keySet().size() != 1) {
+                    logger.error("Invalid bulk-update: more than one type given {}", String.join(", ", updateItem.keySet()));
+                    session.transfer(flowFile, REL_FAILURE);
+                    context.yield();
+                    return;
+                }
+                final String updateType = updateItem.keySet().iterator().next();
+                final BsonDocument updateSpec = (BsonDocument) updateItem.get(updateType);
+                final WriteModel<Document> writeModel;
+                if ("insertOne".equals(updateType)) {
+                    writeModel = new InsertOneModel<>(toBsonDocument((BsonDocument) updateSpec.get("document")));
+                } else if ("updateOne".equals(updateType)) {
+                    final UpdateOptions options = parseUpdateOptions(updateSpec);
+                    writeModel = new UpdateOneModel<>((BsonDocument) updateSpec.get("filter"), (BsonDocument) updateSpec.get("update"), options);
+                } else if ("updateMany".equals(updateType)) {
+                    final UpdateOptions options = parseUpdateOptions(updateSpec);
+                    writeModel = new UpdateManyModel<>((BsonDocument) updateSpec.get("filter"), (BsonDocument) updateSpec.get("update"), options);
+                } else if ("replaceOne".equals(updateType)) {
+                    final ReplaceOptions options = parseReplaceOptions(updateSpec);
+                    writeModel = new ReplaceOneModel<>((BsonDocument) updateSpec.get("filter"),
+                            toBsonDocument((BsonDocument) updateSpec.get("replacement")), options);
+                } else if ("deleteOne".equals(updateType)) {
+                    final DeleteOptions options = parseDeleteOptions(updateSpec);
+                    writeModel = new DeleteOneModel<>((BsonDocument) updateSpec.get("filter"), options);
+                } else  if ("deleteMany".equals(updateType)) {
+                    final DeleteOptions options = parseDeleteOptions(updateSpec);
+                    writeModel = new DeleteManyModel<>((BsonDocument) updateSpec.get("filter"), options);
+                } else {
```

Comment:
It looks like this conditional logic could be broken out to a separate method that returns a `WriteModel`, or `null` on failure. That could make the `onTrigger` method easier to maintain.